### PR TITLE
simplewallet: lock console on inactivity

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Comment by moneromooo-monero** [comment]
I was thinking about adding this:

```
 ________________________________________
/ I locked your Monero wallet to protect \
\ you while you were away.               /
 ----------------------------------------
        \   (__)
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

**Comment by xmrdsc** [comment]
+1 mooo ascii

File: src/simplewallet/simplewallet.cpp [hunk 1]
```diff
@@ -2731,83 +2781,86 @@ simple_wallet::simple_wallet()
   , m_auto_refresh_refreshing(false)
   , m_in_manual_refresh(false)
   , m_current_subaddress_account(0)
+  , m_last_activity_time(time(NULL))
+  , m_locked(false)
+  , m_in_command(false)
 {
   m_cmd_binder.set_handler("start_mining",
```
Comment (DanielBullimore) [inline src/simplewallet/simplewallet.cpp]: can mining even start? both start and stop handler look like they stop mining?

**Comment by moneromooo-monero** [comment]
Now also works when HAVE_READLINE is not set. And a cow.

**Comment by erciccione** [comment]
Thanks for the i18n friendly cow :)

**Comment by moneromooo-monero** [comment]
I'm rewriting that part btw.

**Comment by moneromooo-monero** [comment]
Now works better with languages like Japanese where characters are two 'slots' wide.

**Comment by moneromooo-monero** [comment]
and now with all the generic code in utils so it maybe reused

**Comment by NanoAkron** [comment]
Needs more cow(bell)...

**Comment by moneromooo-monero** [comment]
Squashed and rebased to latest master.

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -3944,6 +3951,8 @@ bool wallet2::load_keys(const std::string& keys_file_name, const epee::wipeable_
     m_ignore_fractional_outputs = field_ignore_fractional_outputs;
     GET_FIELD_FROM_JSON_RETURN_ON_ERROR(json, track_uses, int, Int, false, false);
     m_track_uses = field_track_uses;
+    GET_FIELD_FROM_JSON_RETURN_ON_ERROR(json, inactivity_lock_timeout, unsigned int, Uint, false, DEFAULT_INACTIVITY_LOCK_TIMEOUT);
```
Comment (xiphon) [inline src/wallet/wallet2.cpp]:
```suggestion
    GET_FIELD_FROM_JSON_RETURN_ON_ERROR(json, inactivity_lock_timeout, uint32_t, Uint, false, DEFAULT_INACTIVITY_LOCK_TIMEOUT);
```

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -3315,6 +3376,7 @@ bool simple_wallet::set_variable(const std::vector<std::string> &args)
     CHECK_SIMPLE_VARIABLE("segregation-height", set_segregation_height, tr("unsigned integer"));
     CHECK_SIMPLE_VARIABLE("ignore-fractional-outputs", set_ignore_fractional_outputs, tr("0 or 1"));
     CHECK_SIMPLE_VARIABLE("track-uses", set_track_uses, tr("0 or 1"));
+    CHECK_SIMPLE_VARIABLE("inactivity-lock-timeout", set_inactivity_lock_timeout, tr("unsigned integer"));
```
Comment (xiphon) [inline src/simplewallet/simplewallet.cpp]:
```suggestion
    CHECK_SIMPLE_VARIABLE("inactivity-lock-timeout", set_inactivity_lock_timeout, tr("inactivity timeout in seconds before locking the wallet console, 0 to disable"));
```

File: src/common/util.h [hunk 4]
```diff
@@ -248,4 +248,10 @@ namespace tools
   std::string get_human_readable_timespan(uint64_t seconds);
 
   std::string get_human_readable_bytes(uint64_t bytes);
+
+  void clear_screen();
+
+  std::pair<std::string, size_t> get_string_prefix_by_width(const std::string &s, size_t columns);
```
Comment (xiphon) [inline src/common/util.h]: Can drop `get_string_prefix_by_width` and `get_string_width` from the header.

File: src/simplewallet/simplewallet.cpp [hunk 5]
```diff
@@ -5656,6 +5723,67 @@ bool simple_wallet::prompt_if_old(const std::vector<tools::wallet2::pending_tx>
...
   return true;
 }
+void simple_wallet::check_for_inactivity_lock(bool user)
+{
+  if (m_locked)
+  {
+#ifdef HAVE_READLINE
+    PAUSE_READLINE();
+    rdln::clear_screen();
+#endif
```
Comment (xiphon) [inline src/simplewallet/simplewallet.cpp]:
```suggestion
#else
    tools::clear_screen();
#endif
```

Comment (moneromooo-monero) [reply src/simplewallet/simplewallet.cpp]: Why ? That makes it less likely to work, no ?

Comment (xiphon) [reply src/simplewallet/simplewallet.cpp]: TBH i'm fine with both ways. Feel free to leave it as is if you want.

**Pull request merged**